### PR TITLE
[mlir][vector] Relax the requirements on broadcast dims

### PR DESCRIPTION
**Pull request by banach-space**
NOTE: This is a follow-up for #97049 in which the `in_bounds` attribute
was made mandatory.

This PR updates the semantics of the `in_bounds` attribute so that
broadcast dimensions are no longer required to be "in bounds".
Specifically, these xfer_read/xfer_write Ops become valid after this
change:

```mlir
  %read = vector.transfer_read %A[%base1, %base2], %pad
      {in_bounds = [false], permutation_map = affine_map<(d0, d1) -> (0)>}
      {permutation_map = affine_map<(d0, d1) -> (0)>}
      : memref<?x?xf32>, vector<9xf32>

  vector.transfer_write %vec, %A[%base1, %base2],
      {in_bounds = [false], permutation_map = affine_map<(d0, d1) -> (0)>}
      {permutation_map = affine_map<(d0, d1) -> (0)>}
      : vector<9xf32>, memref<?x?xf32>
```

Note that the value `false` merely means "may run out-of-bounds", i.e.,
the corresponding access can still be "in bounds". In fact, the folder
for xfer Ops is also updated (*) and will update the attribute value
corresponding to broadcast dims to `true` if all non-broadcast dims
are marked as "in bounds". 

Note that this PR doesn't change any of the lowerings. The changes in
"SuperVectorize.cpp", "Vectorization.cpp" and "AffineMap.cpp" are simple
reverts of recent changes in #97049. Those were only meant to facilitate
making `in_bounds` mandatory and to work around the extra requirements
for broadcast dims (those requirements ere removed in this PR). All
changes in tests are also reverts of changes from #97049.

For context, here's a PR in which "broadcast" dims where forced to
always be "in-bounds":
  * https://reviews.llvm.org/D102566

(*) See `foldTransferInBoundsAttribute`.

### CODE REVIEW

**Comment by dcaballe** [comment]
> In fact, the folder
for xfer Ops is also updated (*) and will update the attribute value
corresponding to broadcast dims to true. Indeed, such dims would
never be out-of-bounds in practice.

Could you please elaborate on this? I'm not sure I follow

**Comment by banach-space** [comment]
> Could you please elaborate on this? I'm not sure I follow

I "spoke" too soon :) 

Looking at this example (output from a vectorizer test), we can infer that all broadcast dims are "in bounds" as all other dims **are in bounds**:
```mlir
#map = affine_map<(d0, d1) -> (d0, 0, 0, d1)>
#map1 = affine_map<(d0) -> (d0, 0, 0, 0)>
#map2 = affine_map<(d0) -> (0, 0, d0, 0)>
#map3 = affine_map<(d0, d1) -> (d1, 0, d0, 0)>
module {
  func.func @generic_vectorize_broadcast_transpose(%arg0: memref<4xf32>, %arg1: memref<4x4xf32>, %arg2: memref<4x4x4x4xf32>) {
    %c0 = arith.constant 0 : index
    %cst = arith.constant 0.000000e+00 : f32
    %0 = vector.transfer_read %arg1[%c0, %c0], %cst {in_bounds = [true, false, false, true], permutation_map = #map} : memref<4x4xf32>, vector<4x4x4x4xf32>
    %1 = vector.transfer_read %arg0[%c0], %cst {in_bounds = [true, false, false, false], permutation_map = #map1} : memref<4xf32>, vector<4x4x4x4xf32>
    %2 = vector.transfer_read %arg0[%c0], %cst {in_bounds = [false, false, true, false], permutation_map = #map2} : memref<4xf32>, vector<4x4x4x4xf32>
    %3 = vector.transfer_read %arg1[%c0, %c0], %cst {in_bounds = [true, false, true, false], permutation_map = #map3} : memref<4x4xf32>, vector<4x4x4x4xf32>
    %4 = arith.subf %0, %1 : vector<4x4x4x4xf32>
    %5 = arith.addf %2, %4 : vector<4x4x4x4xf32>
    %6 = arith.addf %3, %5 : vector<4x4x4x4xf32>
    vector.transfer_write %6, %arg2[%c0, %c0, %c0, %c0] {in_bounds = [true, true, true, true]} : vector<4x4x4x4xf32>, memref<4x4x4x4xf32>
    return
  }
```

In cases where non-broadcast dims are not "in bounds", I think that we should conservatively stick to "out of bounds". I will be sending an update for `foldTransferInBoundsAttribute` shortly.

**Comment by banach-space** [comment]
Hey @dcaballe , shall we try finalising this one? Could you take another look?

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 1]
```diff
@@ -4138,22 +4134,42 @@ static LogicalResult foldTransferInBoundsAttribute(TransferOp op) {
...
-    assert(dimExpr && "Broadcast dims must be in-bounds");
-    auto inBounds =
-        isInBounds(op, /*resultIdx=*/i, /*indicesIdx=*/dimExpr.getPosition());
+    if (dimExpr) {
+      // 2.a Non-broadcast dim
+      inBounds = isInBounds(op, /*resultIdx=*/i,
+                            /*indicesIdx=*/dimExpr.getPosition());
+      // 2.b Broadcast dims are handled after processing non-bcast dims
+      // FIXME: constant expr != 0 are not broadcasts - should such
+      // constants be allowed at all?
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:4154]: I don't think so. Any value != 0 would be incorrect.

**Comment by dcaballe** [comment]
Thanks, and sorry for the delay. LGTM

> In cases where non-broadcast dims are not "in bounds", I think that we should conservatively stick to "out of bounds". I will be sending an update for foldTransferInBoundsAttribute shortly.

This is conservative that it should be good enough.

**Review by dcaballe — APPROVED** [review]

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 2]
```diff
@@ -4138,22 +4134,42 @@ static LogicalResult foldTransferInBoundsAttribute(TransferOp op) {
...
   }
+
+  // Handle broadcast dims: if all non-broadcast dims are "in
+  // bounds", then all bcast dims should be "in bounds" as well.
+  bool allNonBcastDimsInBounds = llvm::all_of(
+      nonBcastDims, [&newInBounds](unsigned idx) { return newInBounds[idx]; });
+  if (allNonBcastDimsInBounds)
+    llvm::for_each(permutationMap.getBroadcastDims(), [&](unsigned idx) {
+      changed |= !newInBounds[idx];
+      newInBounds[idx] = true;
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:4170]: nit: why not a plain for loop?

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 1 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:4154]: Thanks! Looks like the Op verifier would reject such cases before we get here (so all good). I'll add relevant "invalid" cases to test/Dialect/Vector/invalid.mlir

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 2 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp]: Updated.

**Comment by banach-space** [comment]
@kuhar, do let me know if you have any further comments, otherwise I will land this tomorrow. Thanks for taking a look 🙏🏻

**Comment by kuhar** [comment]
> @kuhar, do let me know if you have any further comments, otherwise I will land this tomorrow. Thanks for taking a look 🙏🏻

I'm not an expert on broadcasting semantics, no need to wait for me

**Pull request merged**